File: src/executor/plannerClient.ts
import { spawn } from 'node:child_process';
import type { RedactedMessage } from '../redactor.js';
import type { Goal, PlannerInput } from '../planner/index.js';

/**
 * Spawn the planner as a separate child process. The executor's
 * environment is filtered before being handed to the planner so that
 * credential-bearing variables (anything matching CRED / TOKEN / SECRET
 * / KEY / STEWARD_CREDENTIALS_DIR) never reach the LLM-side process.
 *
 * This is the slice-002 enforcement of decision §15 (structural
 * separation). Issue 012 will replace env-based creds with op:// refs.
 */
const FORBIDDEN_ENV = /^(.*(CRED|TOKEN|SECRET|KEY|PASSWORD).*|STEWARD_CREDENTIALS_DIR)$/i;

export function sanitiseEnvForPlanner(
  env: NodeJS.ProcessEnv = process.env,
): NodeJS.ProcessEnv {
  const out: NodeJS.ProcessEnv = {};
  for (const [k, v] of Object.entries(env)) {
    if (FORBIDDEN_ENV.test(k)) continue;
    out[k] = v;
  }
  return out;
}

export async function runPlanner(
  input: PlannerInput | RedactedMessage,
  plannerScript: string,
): Promise<Goal> {
  return new Promise((resolve, reject) => {
    const child = spawn('npx', ['tsx', plannerScript], {
      env: sanitiseEnvForPlanner(),
      stdio: ['pipe', 'pipe', 'inherit'],
    });
    let buf = '';
    child.stdout.on('data', (chunk: Buffer) => {
      buf += chunk.toString('utf8');
    });
    child.on('error', reject);
    child.on('close', () => {
      const line = buf.trim().split('\n').pop() ?? '';
      try {
        const parsed = JSON.parse(line) as Goal | { error: string };
        if ('error' in parsed) reject(new Error(parsed.error));
        else resolve(parsed);
      } catch (err) {
        reject(new Error(`planner output not JSON: ${line}`));
      }
    });
    child.stdin.write(JSON.stringify(input) + '\n');
    child.stdin.end();
  });
}

File: python/steward/executor/planner_client.py
"""Spawn the planner as a subprocess. Sanitise credential-bearing env vars before exec."""
from __future__ import annotations

import asyncio
import json
import os
import re
from typing import Any

from steward.planner import Goal

FORBIDDEN_ENV = re.compile(r"^(.*(CRED|TOKEN|SECRET|KEY|PASSWORD).*|STEWARD_CREDENTIALS_DIR)$", re.IGNORECASE)


def sanitise_env_for_planner(env: dict[str, str] | None = None) -> dict[str, str]:
    source = env if env is not None else dict(os.environ)
    return {k: v for k, v in source.items() if not FORBIDDEN_ENV.match(k)}


async def run_planner(input_data: dict[str, Any]) -> Goal:
    proc = await asyncio.create_subprocess_exec(
        "python",
        "-m",
        "steward.planner",
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        env=sanitise_env_for_planner(),
    )
    stdout, stderr = await proc.communicate(
        (json.dumps(input_data) + "\n").encode("utf-8")
    )
    line = stdout.decode("utf-8").strip().split("\n")[-1] if stdout else ""
    try:
        parsed = json.loads(line)
    except json.JSONDecodeError as e:
        raise RuntimeError(f"planner output not JSON: {line!r}") from e
    if "error" in parsed:
        raise RuntimeError(parsed["error"])
    return Goal(**parsed)
